aten/src/ATen/core/type.cpp

```diff
@@ -702,6 +716,174 @@ std::string TupleType::python_str_impl(TypePrinter printer) const {
 }
 
+static std::vector<bool> findContiguous(
+    const at::IntArrayRef& sizes,
+    const at::IntArrayRef& strides) {
+  AT_ASSERT(sizes.size() == strides.size());
+  std::vector<bool> cont(sizes.size());
+  for (size_t i = 0; i < sizes.size(); ++i) {
+    const auto expected_stride =
+        (i + 1 < sizes.size()) ? sizes[i + 1] * strides[i + 1] : 1;
+    cont[i] = (strides[i] == expected_stride);
+  }
+  return cont;
+}
+
+
+  VaryingShape<int64_t> TensorType::strides() const {
+    if (!strides_.size().has_value()) {
+      return VaryingShape<int64_t>();
+    }
+    std::vector<c10::optional<int64_t>> ss(*strides_.size());
+    for (size_t i = 0; i < *strides_.size(); i++) {
+      if (!strides_[i].has_value()) {
+        continue;
+      }
+      auto s = *strides_[i];
+      if (s.stride_index_.has_value() && s.stride_.has_value()) {
+        ss[*s.stride_index_] = *s.stride_;
+      }
+    }
+    return VaryingShape<int64_t>(ss);
+  }
+
+VaryingShape<Stride> TensorType::
+    computeStrideProps(at::IntArrayRef sizes, at::IntArrayRef strides) {
+
+  std::vector<size_t> stride_indices(sizes.size());
+  std::iota(stride_indices.begin(), stride_indices.end(), 0);
+
+  std::sort(
+      stride_indices.begin(),
+      stride_indices.end(),
+      [&strides](const int& a, const int& b) {
+        // break ties in case of unsqueezed dims
+        // i.e. (1, 1, 5)
+        if (strides[a] == strides[b]) {
+          return a > b;
+        }
+        return strides[a] < strides[b];
+      });
+
+  std::vector<Stride> stride_props;
+  for (size_t i = 0; i < stride_indices.size(); i++) {
+    Stride s {stride_indices[i], false, strides[stride_indices[i]]};
+    // innermost stride expected to be 1
+    // TODO: turn contiguous_ into an enum CONTIGUOUS, NONCONTIGUOUS, BROADCASTED
+    if (stride_indices[i] == stride_indices.size() - 1) {
+      s.contiguous_ = strides[stride_indices[i]] == 1;
+    } else {
+      s.contiguous_ = strides[stride_indices[i]] == strides[stride_indices[i-1]]*sizes[stride_indices[i-1]] && strides[stride_indices[i]] != 0;
+    }
+    stride_props.push_back(s);
+  }
+
+  return VaryingShape<Stride>{stride_props};
+}
+
+  std::atomic<size_t> ShapeSymbol::num_symbols {1};
+
+  template struct VaryingShape<c10::ShapeSymbol>;
+  template struct VaryingShape<bool>;
+  template struct VaryingShape<size_t>;
+  template struct VaryingShape<int64_t>;
+
+
+  TensorType::TensorType(
+      c10::optional<at::ScalarType> scalar_type,
+      c10::optional<Device> device,
+      const VaryingShape<ShapeSymbol>& sizes,
+      const VaryingShape<Stride>& strides,
+      c10::optional<bool> requires_grad,
+      c10::optional<bool> undefined)
+      : Type(TypeKind::TensorType),
+        scalar_type_(scalar_type),
+        device_(device),
+        sizes_(sizes),
+        strides_(strides),
+        requires_grad_(requires_grad),
+        undefined_(undefined) {}
+
+
+  TensorTypePtr TensorType::create(const at::Tensor& t) {
```

Comment:
stride_indices, contiguity are unused here. Doesn't seem to be implemented correctly but the code makes it look like it is.

Reply:
oh :-( yes, that's definitely a bug. we should use `computeStrideProps`

Reply:
oh wait, no it's correct. that particular constructor we call is a convenience c-tor. If `VaryingShape`s contain concrete sizes we will compute the stride properties, otherwise we will just compute the rank information.